[MathML] Layout invalid markup as an <mrow> https://bugs.webkit.org/show_bug.cgi?id=123348

icesfont opened the pull request:
#### eaf4e99ac1b02eb53f8b2f157a4c5bbec7cf2b68
<pre>
[MathML] Layout invalid markup as an &lt;mrow&gt; <a href="https://bugs.webkit.org/show_bug.cgi?id=123348">https://bugs.webkit.org/show_bug.cgi?id=123348</a>

Reviewed by @fred-wang.

Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/ad94cb830eef8b8ff67a41820b3a61519d6d2dbd">https://github.com/web-platform-tests/wpt/commit/ad94cb830eef8b8ff67a41820b3a61519d6d2dbd</a>

Change to make invalid MathML markup lay out as `mrow` by calling `RenderMathMLRow` methods when invalid. Additionally, make elements inheriting from `RenderMathMLRow` and `MathMLRowElement` explicitly override the methods that `RenderMathMLRow`/`MathMLRowElement` override in order to prevent elements erroneously behaving like `mrow`s when valid.

Remove the now unused `RenderMathMLBlock::layoutInvalidMarkup`.

Amend the check for validity in `RenderMathMLScripts::validateAndGetReferenceChildren` to match the spec [1].

Important relevant `mrow` fallback tests from WPT and update test expectations.

[1]: <a href="https://w3c.github.io/mathml-core/spec.html#prescripts-and-tensor-indices-mmultiscripts#">https://w3c.github.io/mathml-core/spec.html#prescripts-and-tensor-indices-mmultiscripts#</a>:~:text=that%20is%20not%20an%20mprescripts%20element
Canonical link: <a href="https://commits.webkit.org/281218@main">https://commits.webkit.org/281218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d036e1f40f74f19dc765c2755910990cd55200de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59091 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38419 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62945 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9534 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61220 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9740 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47985 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6816 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35940 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51142 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28821 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32663 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8407 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8538 "Hash d036e1f4 for PR 30705 does not build (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54618 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8688 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64634 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3003 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8644 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55300 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3014 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51144 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55416 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13100 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2526 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34248 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35332 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36417 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35078 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/mathml/MathMLFractionElement.cpp:
```diff
@@ -42,7 +42,7 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(MathMLFractionElement);
 
 inline MathMLFractionElement::MathMLFractionElement(const QualifiedName& tagName, Document& document)
-    : MathMLPresentationElement(tagName, document)
+    : MathMLRowElement(tagName, document)
```
I wonder whether we could avoid that, but anyway MathMLRowElement::childrenChanged overrides the following methods:

```
    void childrenChanged(const ChildChange&) override;
    bool acceptsMathVariantAttribute() override;
    RenderPtr<RenderElement> createElementRenderer(RenderStyle&&, const RenderTreePosition&) override;
```

* MathMLRowElement::createElementRenderer is overridden in MathMLFractionElement/MathMLScriptsElement and don't call MathMLPresentationElement's implementation, so there is no behavior change.

* MathMLRowElement::acceptsMathVariantAttribute returns true if the element is an mstyle tag. It's ok because it will be false for MathMLFractionElement/MathMLScriptsElement and hasTagName comparison should be fast. But maybe better to override it explicitly to return false.

* MathMLRowElement::childrenChanged do some invalidation for mo children, when the child list changed and is not overridden in MathMLFractionElement/MathMLScriptsElement. I believe that would not change the behavior so that's probably fine as is, but maybe just add a comment in MathMLRowElement::childrenChanged that it would be good to avoid this invalidation for MathMLFractionElement/MathMLScriptsElement if renderer() is non-null and returns "is valid".

icesfont replied:
Thank you for the comments! I've overridden MathMLRowElement::acceptsMathVariantAttribute explicitly in MathMLFractionElement and MathMLScriptsElement. For the 3rd bullet point, I'm not sure what renderer() refers to, please could you elaborate?

fred-wang replied:
Yes, so the `Element` objects (DOM) create their associated `RenderElement` object (layout & painting) via `createElementRenderer()`: https://searchfox.org/wubkat/rev/7eea0493a7279f9a03309c5f867ee8e947ca9cdd/Source/WebCore/dom/Element.h#392

Once it's created, that the `renderer()` method will return a non-null pointer: https://searchfox.org/wubkat/rev/7eea0493a7279f9a03309c5f867ee8e947ca9cdd/Source/WebCore/dom/Element.h#392

So if `RenderMathMLFraction::isValid()` becomes public, `MathMLFractionElement::childrenChanged` could do something like:

```
  if (renderer() && !downcast<RenderMathMLFraction>(*renderer()).isValid()) {
     MathMLRowElement::childrenChanged(change);
     return;
  }
  MathMLPresentationElement::childrenChanged(change);
```

icesfont replied:
I see, thanks a lot! Added the comment.

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
```diff
@@ -42,7 +42,7 @@ namespace WebCore {
 
 RenderMathMLFraction::RenderMathMLFraction(MathMLFractionElement& element, RenderStyle&& style)
-    : RenderMathMLBlock(Type::MathMLFraction, element, WTFMove(style))
+    : RenderMathMLRow(Type::MathMLFraction, element, WTFMove(style))
```
RenderMathMLRow overrides the following methods:


```
    void layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalHeight = 0_lu) override;
    std::optional<LayoutUnit> firstLineBaseline() const override;
    void computePreferredLogicalWidths() override;
    ASCIILiteral renderName() const override { return "RenderMathMLRow"_s; }
 ```

* If I'm correct, there are overridden methods in RenderMathMLFraction/RenderMathMLRoot/RenderMathMLScripts, except RenderMathMLRoot::firstLineBaseline(), please add one.

* I believe renderName() never calls the method of the parent class, so nothing to do.

* For layoutBlock(), firstLineBaseline() and computePreferredLogicalWidths() you should make sure that we call RenderMathMLRow's method if the renderer is not valid and run the class's implementation otherwise.

icesfont replied:
Added RenderMathMLRoot::firstLineBaseline(). 3rd bullet point *should* be enforced

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
```diff
@@ -229,7 +229,7 @@ void RenderMathMLFraction::layoutBlock(bool relayoutChildren, LayoutUnit)
 
     if (!isValid()) {
-        layoutInvalidMarkup(relayoutChildren);
+        RenderMathMLRow::layoutBlock(relayoutChildren);
```
IIUC layoutInvalidMarkup is not used anymore, so please just it from RenderMathMLBlock.cpp/h

icesfont replied:
Yep

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
```diff
@@ -125,6 +125,10 @@ std::optional<RenderMathMLScripts::ReferenceChildren> RenderMathMLScripts::valid
+        // The base cannot be an <mprescripts> element.
+        if (isPrescriptDelimiter(*base))
+            return std::nullopt;
+
```
Please remember to mention this change in you commit message too.

icesfont replied:
Alright, mentioned

fred-wang reviewed (changes requested):
Thanks Harry. That looks a good start!

I left some inline comments but in general:
* I think you'll need more changes (and to update your commit message later) 
* You'll also need to check the failing MathML tests, probably just rebaseline update the expectations.
* mrow-fallback.html is only checking layout. We need to add WPT tests for preferred width and baseline too (let's discuss that on WPT github).

fred-wang commented on Source/WebCore/mathml/MathMLFractionElement.cpp:
```diff
@@ -135,6 +135,11 @@ RenderPtr<RenderElement> MathMLFractionElement::createElementRenderer(RenderStyl
+{
+    return false;
+}
+
```
This is a small function that can be inlined by the compiler, please just move it to the header as in https://searchfox.org/wubkat/rev/7eea0493a7279f9a03309c5f867ee8e947ca9cdd/Source/WebCore/mathml/MathMLTokenElement.h#54

icesfont replied:
Yep

fred-wang commented on Source/WebCore/mathml/MathMLFractionElement.h:
```diff
@@ -47,6 +47,7 @@ class MathMLFractionElement final : public MathMLPresentationElement {
 private:
     MathMLFractionElement(const QualifiedName& tagName, Document&);
     RenderPtr<RenderElement> createElementRenderer(RenderStyle&&, const RenderTreePosition&) final;
+    bool acceptsMathVariantAttribute() final;
```
just do `{ return false; }` here

fred-wang commented on Source/WebCore/mathml/MathMLScriptsElement.cpp:
```diff
@@ -93,6 +93,11 @@ RenderPtr<RenderElement> MathMLScriptsElement::createElementRenderer(RenderStyle
+{
+    return false;
+}
+
```
ditto

fred-wang commented on Source/WebCore/mathml/MathMLScriptsElement.h:
```diff
@@ -46,6 +46,7 @@ class MathMLScriptsElement : public MathMLPresentationElement {
 
 private:
     RenderPtr<RenderElement> createElementRenderer(RenderStyle&&, const RenderTreePosition&) override;
+    bool acceptsMathVariantAttribute() override;
```
ditto

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
```diff
@@ -481,9 +485,11 @@ void RenderMathMLScripts::layoutBlock(bool relayoutChildren, LayoutUnit)
 std::optional<LayoutUnit> RenderMathMLScripts::firstLineBaseline() const
 {
+    auto possibleReference = validateAndGetReferenceChildren();
+    if (!possibleReference)
```
`possibleReference` is not used in this function, so that could just be

`if (!validateAndGetReferenceChildren())`

but maybe better to do

`auto& base = *possibleReference.value().base;` 

since the validation ensured we have a base.

(It's a pity that mmultiscripts validation has linear complexity rather than constant, but I probably that's fine since the element is rare and generally only have a few children.)

icesfont replied:
Changed to this

icesfont replied:
Changed to the latter

fred-wang replied:
Thanks

fred-wang reviewed (changes requested):
Thank you that's getting closer!

Note: there are two empty `` in the commit message.

As we discussed, you'll also need to update the test expectations and import the new tests from upstream WPT.

fred-wang commented on Source/WebCore/mathml/MathMLRowElement.cpp:
```diff
@@ -54,6 +54,9 @@ Ref<MathMLRowElement> MathMLRowElement::create(const QualifiedName& tagName, Doc
 
 void MathMLRowElement::childrenChanged(const ChildChange& change)
 {
+    // This does not change the behaviour for valid MathMLFractionElement/MathMLScriptsElement, which inherit this.
```
nit: it might be observable if that makes the code slower.

Anyway, can you please just open a bug in bugzilla where you put the details, and just make this a a FIXME like this:

// FIXME: Avoid this invalidation for valid MathMLFractionElement/MathMLScriptsElement  (http://wkbug/XXXXX)

icesfont replied:
Added

fred-wang reviewed (commented):
It seems presentation-markup/mrow has quite a few tests that we don't have imported yet. Can you please handle the import in a separate PR (with its own bug)? That will make the easier to review the patches

fred-wang commented on Source/WebCore/mathml/MathMLRowElement.cpp:
```diff
@@ -54,6 +54,7 @@ Ref<MathMLRowElement> MathMLRowElement::create(const QualifiedName& tagName, Doc
 
 void MathMLRowElement::childrenChanged(const ChildChange& change)
 {
+    // FIXME: Avoid this invalidation for valid MathMLFractionElement/MathMLScriptsElement (http://wkbug/276828)
```
Need to use periods, see https://webkit.org/code-style-guidelines/#comments-sentences

Also this shortcut URL doesn't work anymore?. Maybe just add a new sentence on the a new line "See https://bugs.webkit.org/show_bug.cgi?id=276828."

icesfont replied:
Added

fred-wang commented on LayoutTests/TestExpectations:
```diff
@@ -1790,6 +1790,7 @@ imported/w3c/web-platform-tests/html/browsers/windows/nested-browsing-contexts/f
 webkit.org/b/180013 mathml/non-core/lengths-2.html [ ImageOnlyFailure ]
 webkit.org/b/194952 imported/w3c/web-platform-tests/mathml/presentation-markup/fractions/frac-bar-001.html [ ImageOnlyFailure ]
 webkit.org/b/187039 imported/w3c/web-platform-tests/mathml/relations/html5-tree/href-click-3.html [ Skip ]
+webkit.org/b/123348 mathml/radical-fallback.html [ ImageOnlyFailure ]
```
OK https://searchfox.org/wubkat/source/LayoutTests/mathml/radical-fallback.html is really a pixel test, so I guess we should update the PNG expectations here: https://searchfox.org/wubkat/search?q=radical-fallback&path=&case=true&regexp=false

But see below.

icesfont replied:
Got it -- I'll remove this then

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
```diff
@@ -313,6 +312,14 @@ void RenderMathMLRoot::paint(PaintInfo& info, const LayoutPoint& paintOffset)
+    if (!isValid())
+        return RenderMathMLRow::firstLineBaseline();
+
+    return RenderMathMLBlock::firstLineBaseline();
```
So this what is causing a behavior change, at least visible for `radical-fallback.html`.

Before this patch `RenderMathMLRoot::firstLineBaseline()` did not exist but `RenderMathMLRoot` was already inheriting from `RenderMathMLRow` (contrary to fraction/scripts). Now, invalid mroot markup continues to use `RenderMathMLRow::firstLineBaseline()` but msqrt (always valid) and mroot will use `RenderMathMLBlock::firstLineBaseline()` instead.

Checking the two methods, I believe keep `RenderMathMLRow::firstLineBaseline()` (using the baseline of the first child as the baseline of the whole msqrt/mroot) is also correct (see links below). So we don't need to override `firstLineBaseline()` here.

> https://w3c.github.io/mathml-core/#figure-box-msqrt
> The [alphabetic baseline](https://w3c.github.io/mathml-core/#dfn-alphabetic-baseline) of the [msqrt base](https://w3c.github.io/mathml-core/#dfn-msqrt-base) is aligned with the [alphabetic baseline](https://w3c.github.io/mathml-core/#dfn-alphabetic-baseline).

> https://w3c.github.io/mathml-core/#figure-box-mroot
> The [alphabetic baseline](https://w3c.github.io/mathml-core/#dfn-alphabetic-baseline) of B is aligned with the [alphabetic baseline](https://w3c.github.io/mathml-core/#dfn-alphabetic-baseline).

> https://w3c.github.io/mathml-core/#layout-of-mrow
> For each [in-flow](https://www.w3.org/TR/css-display-3/#in-flow) child ... its [block offset](https://w3c.github.io/mathml-core/#dfn-block-offset) such that the [alphabetic baseline](https://w3c.github.io/mathml-core/#dfn-alphabetic-baseline) of the child is aligned with the [alphabetic baseline](https://w3c.github.io/mathml-core/#dfn-alphabetic-baseline).

icesfont replied:
I see, thank you!

fred-wang reviewed (changes requested):
Can you please rebase this PR, so we have only the second commit?

fred-wang commented on LayoutTests/mathml/radical-fallback-expected.txt:
```diff
@@ -0,0 +1,23 @@
+layer at (0,0) size 800x600
```
I understand this new file is not necessary since (1) we have platform-specific expectations for win, mac, ios and glib already. (2) you removed the changed to `RenderMathMLRoot::firstLineBaseline`

icesfont replied:
Deleted

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLUnderOver.cpp:
```diff
@@ -380,6 +379,14 @@ void RenderMathMLUnderOver::layoutBlock(bool relayoutChildren, LayoutUnit pageLo
+
+    return RenderMathMLScripts::firstLineBaseline();
+}
+
```
OK, I guess this one is not necessary either, since  RenderMathMLScripts::firstLineBaseline performs a validation now...

icesfont replied:
Oops, yep -- I'd thought that it would need its own validation, but looking more closely this validation is really just a subset of the validation done in RenderMathMLScripts anyway

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLScripts.h:
```diff
@@ -50,17 +50,17 @@ class RenderMathMLScripts : public RenderMathMLBlock {
     MathMLScriptsElement::ScriptType scriptType() const;
     void computePreferredLogicalWidths() override;
     void layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalHeight = 0_lu) override;
+    std::optional<LayoutUnit> firstLineBaseline() const override;
```
I guess you did that to use it in RenderMathMLUnderOver? For that, you only need it to be `protected`, but I guess you don't need to override RenderMathMLUnderOver::firstLineBaseline anyway (see below)...

fred-wang replied:
also I believe you can keep final too

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
```diff
@@ -68,7 +68,7 @@ RenderMathMLOperator* RenderMathMLScripts::unembellishedOperator() const
     return base ? base->unembellishedOperator() : nullptr;
```
Can you please do like in `RenderMathMLScripts::firstLineBaseline()` and call the RenderMathMLRow's version or the base's unembellishedOperator ?

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
```diff
@@ -181,15 +181,15 @@ void RenderMathMLFraction::computePreferredLogicalWidths()
 {
```
For  RenderMathMLFraction::unembellishedOperator(), can you please use `RenderMathMLRow::unembellishedOperator()` when  not valid? That won't change behavior for now as we don't implement what's in the spec and just return nullptr, but good to be explicit.

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLBlock.h:
```diff
@@ -77,7 +77,6 @@ class RenderMathMLBlock : public RenderBlock {
     static inline LayoutUnit ascentForChild(const RenderBox& child);
```
can you please use nullptr instead of 0 in unembellishedOperator()  ?

fred-wang reviewed (changes requested):
This LGTM, but I have more inline comments....

The new debug assertion might be appearing because we have an invalid munderover markup here: https://searchfox.org/wubkat/rev/3dea3f050e72b74b587cf5f4943a60c7b5cb5452/LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/stretch-along-block-axis-001.html#154

(you can see an console error if you open http://wpt.live/mathml/presentation-markup/mrow/stretch-along-block-axis-001.html in Firefox)

In the past layoutAsInvalidMarkup would not stretch the operator, but with your patch we just  use RenderMathMLRow which does stretch the operator.

Ultimately we need to fix the test in WPT, but let's do that in a follow-up. For now, you can probably just mark the test as "Crash" or "Crash Pass" in the TestExpectations file.

It's also possible that the change to RenderMathMLScripts::unembellishedOperator could help to work around the Crash and so changing the TestExpectations won't be necessary to land this patch...

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
```diff
@@ -64,11 +64,14 @@ MathMLScriptsElement::ScriptType RenderMathMLScripts::scriptType() const
+    if (!possibleReference)
+        return RenderMathMLRow::unembellishedOperator();
+
+    return dynamicDowncast<RenderMathMLBlock>(*possibleReference.value().base)->unembellishedOperator();
```
I think you need to keep a RenderMathMLBlock check like

```
auto* base = dynamicDowncast<RenderMathMLBlock>(possibleReference.value().base);
return base ? base->unembellishedOperator() : nullptr;
```

icesfont committed:
[MathML] Layout invalid markup as an <mrow> https://bugs.webkit.org/show_bug.cgi?id=123348

Reviewed by @fred-wang.

Upstream commit: https://github.com/web-platform-tests/wpt/commit/ad94cb830eef8b8ff67a41820b3a61519d6d2dbd

Change to make invalid MathML markup lay out as `mrow` by calling `RenderMathMLRow` methods when invalid. Additionally, make elements inheriting from `RenderMathMLRow` and `MathMLRowElement` explicitly override the methods that `RenderMathMLRow`/`MathMLRowElement` override in order to prevent elements erroneously behaving like `mrow`s when valid.

Remove the now unused `RenderMathMLBlock::layoutInvalidMarkup`.

Amend the check for validity in `RenderMathMLScripts::validateAndGetReferenceChildren` to match the spec [1].

Important relevant `mrow` fallback tests from WPT and update test expectations.

[1]: https://w3c.github.io/mathml-core/spec.html#prescripts-and-tensor-indices-mmultiscripts#:~:text=that%20is%20not%20an%20mprescripts%20element
Canonical link: https://commits.webkit.org/281218@main
